apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java

```diff
@@ -579,22 +590,79 @@ public long countBlobs(String containerName, ListContainerOptions options) {
       throw new UnsupportedOperationException();
    }
 
-   // copied from BaseBlobStore
+   // TODO: Move statics up to BaseBlobStore to reuse?
+   @com.google.inject.Inject(optional = true)
+   @Named("jclouds.mpu.parallel.degree")
+   @VisibleForTesting
+   private int numThreads = 16;
+
+   private ExecutorService jcloudsBigBlobUploadTreadPool = MoreExecutors.getExitingExecutorService(
+         new ThreadPoolExecutor(
+         numThreads,
+         numThreads,
+         0L,
+         TimeUnit.MILLISECONDS,
+         new ArrayBlockingQueue<Runnable>(numThreads),
+         new ThreadPoolExecutor.CallerRunsPolicy()), 0, TimeUnit.NANOSECONDS);
+
    @Beta
    protected String putMultipartBlob(String container, Blob blob, PutOptions overrides) {
-      List<MultipartPart> parts = Lists.newArrayList();
+      if (overrides.getCustomExecutor() != null) {
+         return putMultipartBlob(container, blob, overrides, overrides.getCustomExecutor());
+      } else {
+         return putMultipartBlob(container, blob, overrides, jcloudsBigBlobUploadTreadPool);
+      }
+   }
+
+   // copied from BaseBlobStore
+   @Beta
+   protected String putMultipartBlob(String container, Blob blob, PutOptions overrides, ExecutorService executor) {
+      ArrayList<Future<MultipartPart>> parts = new ArrayList<Future<MultipartPart>>();
+
       long contentLength = checkNotNull(blob.getMetadata().getContentMetadata().getContentLength(),
             "must provide content-length to use multi-part upload");
       MultipartUploadSlicingAlgorithm algorithm = new MultipartUploadSlicingAlgorithm(
             getMinimumMultipartPartSize(), getMaximumMultipartPartSize(), getMaximumNumberOfParts());
       long partSize = algorithm.calculateChunkSize(contentLength);
       MultipartUpload mpu = initiateMultipartUpload(container, blob.getMetadata(), partSize, overrides);
       int partNumber = 1;
+
       for (Payload payload : slicer.slice(blob.getPayload(), partSize)) {
-         MultipartPart part = uploadMultipartPart(mpu, partNumber, payload);
-         parts.add(part);
-         ++partNumber;
+         BlobUploader b =
+               new BlobUploader(mpu, partNumber++, payload);
+         parts.add(executor.submit(b));
+      }
+
+      return completeMultipartUpload(mpu, Lists.newArrayList(Iterables.transform(parts, new Function<Future<MultipartPart>, MultipartPart>() {
+         @Override
+         public MultipartPart apply(Future<MultipartPart> input) {
```

Comment:
Call `Futures.allAsList().get()` instead.  Possibly with `Futures.getUnchecked` to handle exceptions.